banded Smith-Waterman

pmarks opened the pull request:
Not 100% ready to merge, but I wanted to start the discussion. This build on the sparse-dp code to provide a banded SW.  See header for details.

johanneskoester commented on src/alignment/mod.rs:
```diff
@@ -216,7 +217,66 @@ impl Alignment {
         }
 
-        format!("{}\n{}\n{}", x_pretty, inb_pretty, y_pretty)
+        let mut s = String::new();
```
Can you provide a motivation for this block in contrast to the one-liner before. I guess it formats differently, but ffrom the code I just cannot see immediately from the code.

johanneskoester commented on src/data_structures/bit_tree.rs:
```diff
@@ -0,0 +1,105 @@
+/// to the bit tree with set(j, e), where j <= i. Initially all positions have
+/// the value T::default(). Note that a set cannot be 'undone' by inserting
+/// a smaller element at the same index.
+pub struct MaxBitTree<T: Default + Ord> {
```
The Wikipedia article is named Fenwick Tree, so maybe this should rather be called `FenwickTree`?

johanneskoester commented on src/data_structures/bit_tree.rs:
```diff
@@ -0,0 +1,105 @@
+        let mut idx = idx + 1;
+        let mut sum = T::default();
+        while idx > 0 {
+            sum = max(sum, self.tree[idx].clone());
```
I guess the code would be the same for a prefix-sum? Can this be generalized into a trait `FenwickTree` trait, and you derive `MaxFenwickTree` and `SumFenwickTree`, only implementing a method `operation(T, T)`?

johanneskoester commented on src/alignment/sparse.rs:
```diff
@@ -0,0 +1,452 @@
+    let slc1 = seq1.as_ref();
+    let slc2 = seq2.as_ref();
+
+    let mut set: HashMap<&[u8], Vec<u32>> = HashMap::new();
```
Ok, not using qgramindex is totally fine for one-time use. However, I wonder if it would still be beneficial to iterate over and store bit-encoded qgrams instead of byte slices. But this can also be checked later if you like. Also, can you rename everything from kmer to qgram? Just to be consistent with the rest of the library.

johanneskoester commented on src/alignment/sparse.rs:
```diff
@@ -0,0 +1,452 @@
+
+
+#[cfg(test)]
+mod sparse_alignment {
```
Very nice tests. Great job!

johanneskoester commented on src/data_structures/qgram_index.rs:
```diff
@@ -108,6 +108,11 @@ impl QGramIndex {
         &self.pos[self.address[qgram]..self.address[qgram + 1]]
     }
 
+    /// Iterate over the q-gram ranks of pattern
```
I don't think this method belongs into the qgram index. As you don't need it anyways, would it be ok to remove it?

johanneskoester commented on src/alignment/mod.rs:
```diff
@@ -99,36 +103,94 @@ impl Alignment {
+
+    /// Calculate the bam cigar from the alignment struct. x is the target string
+    /// and y is the reference.
+    pub fn bam_cigar(&self, hard_clip: bool) -> Vec<bam::record::Cigar> {
```
I would prefer to have an `impl From<Alignment> for bam::record::Cigar` in rust-htslib.

pmarks replied:
agreed - moving these methods to rust-htslib

johanneskoester commented on src/data_structures/bit_tree.rs:
```diff
@@ -0,0 +1,105 @@
+// Copyright 2014-2015 Johannes Köster, Vadim Nazarov, Patrick Marks
```
Has Vadim contributed to this? I think it should be you and your colleagues only, Patrick. Not even me until I'll ever touch this code.

pmarks commented:
@johanneskoester  - OK, I've addressed all the comments above.  This has been tested fairly heavliy by some a number of our use cases so I think it's in good shape to merge.

johanneskoester commented:
This is excellent work! Thank you so much! Any known performance changes compared to the previous implementation?

pmarks commented:
Here are the benchmarks, with a couple different string lengths. On shorter strings the timing is about the same, on longer strings it looks like we're a bit slower (but of course that's where banded alignment is critical). @sreenathkrishnan does this look like what you were seeing?

800bp strings:
New Code:
test bench_aligner_wc_global     ... bench:   9,594,821 ns/iter (+/- 28,374,903)
test bench_aligner_wc_local      ... bench:  10,130,867 ns/iter (+/- 24,304,697)
test bench_aligner_wc_semiglobal ... bench:  11,361,908 ns/iter (+/- 2,057,837)

Previous Code:
test bench_aligner_wc_global     ... bench:   9,922,108 ns/iter (+/- 7,385,466)
test bench_aligner_wc_local      ... bench:  12,980,271 ns/iter (+/- 4,386,527)
test bench_aligner_wc_semiglobal ... bench:   9,669,481 ns/iter (+/- 3,372,422)

1500bp strings:
Current code:
test bench_aligner_wc_global     ... bench:  31,048,651 ns/iter (+/- 4,130,667)
test bench_aligner_wc_local      ... bench:  35,190,939 ns/iter (+/- 3,769,121)
test bench_aligner_wc_semiglobal ... bench:  36,942,448 ns/iter (+/- 5,728,201)

Previous master:
test bench_aligner_wc_global     ... bench:  26,460,075 ns/iter (+/- 2,396,104)
test bench_aligner_wc_local      ... bench:  37,115,245 ns/iter (+/- 3,785,367)
test bench_aligner_wc_semiglobal ... bench:  28,529,371 ns/iter (+/- 2,917,783)

sreenathkrishnan commented:
In the cases, I had benchmarked (500 to 5000bp), the relative performance (Time for the new code/Time for the previous master) varied between 0.8 to 1.2, with the new code being about 6% slower on average. These results look pretty much in that ballpark.

pmarks commented:
The rust simd library is pretty far along:
https://github.com/rust-lang-nursery/stdsimd

At some point we'll take a crack at an AVX2 version of this code.
Sreenath's restructuring in this PR should make that a lot easier.


On Tue, Dec 12, 2017 at 8:32 AM sreenathkrishnan <notifications@github.com>
wrote:

> In the cases, I had benchmarked (500 to 5000bp), the relative performance
> (Time for the new code/Time for the previous master) varied between 0.8 to
> 1.2, with the new code being about 6% slower on average. These results look
> pretty much in that ballpark.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-bio/rust-bio/pull/104#issuecomment-351106519>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACzA_4gHS9AdQkG5BL_VjAhxhd0NbKWks5s_qqLgaJpZM4LDpEc>
> .
>

johanneskoester commented:
Hey guys,
one question: was there a particular reason for adding `Xclip` and `Yclip` operations although we already encoded start and end positions in the `Alignment` struct? This seems redundant to me.

sreenathkrishnan commented:
The `Xclip` and `Yclip` operations do not encode any additional information other than the start/end positions of Alignment.  For the standard alignment modes (Global, Semiglobal, and Local), those operations are not included in the `operations` field in `Alignment` struct. 

However, for the custom alignment mode, I would like to think of clipping as a fundamental operation similar to Match, Deletion, etc.; hence I have added them to the `AlignmentOperations` struct, and they are included in the `operations` field in `Alignment` struct.
